File: CS/WebSite/App_Code/DateTemplate.cs
using System;
using System.Data;
using System.Configuration;
using System.Web;
using System.Web.Security;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;
using DevExpress.Web;


public class DataTemplate : ITemplate {
    AccessDataSource accessDS;

    public DataTemplate(AccessDataSource accessDS) {
        this.accessDS = accessDS;
    }

    public void InstantiateIn(Control container) {
        ASPxDropDownEdit dde = container.NamingContainer.NamingContainer as ASPxDropDownEdit;
        string colName = dde.ClientInstanceName;

        ASPxListBox listBox = new ASPxListBox();
        container.Controls.Add(listBox);

        listBox.ID = "listBox";
        listBox.Width = new Unit(100, UnitType.Percentage);
        listBox.ClientInstanceName = "cls" + colName;
        listBox.SelectionMode = ListEditSelectionMode.CheckColumn;
        listBox.ClientSideEvents.SelectedIndexChanged = string.Format("function(s,e) {{ OnListBoxSelectionChanged(s,e,{0}); }}", colName);

        string command = string.Format("SELECT DISTINCT [{0}] FROM [Categories]", colName);
        AccessDataSource ds = new AccessDataSource(accessDS.DataFile, command);

        DataView dv = ds.Select(DataSourceSelectArguments.Empty) as DataView;
        listBox.Items.Add("(Select all)");
        for (int i = 0; i < dv.Count; i++) {
            listBox.Items.Add(dv[i][0].ToString(),i );
        }

        Table table = new Table();
        table.Width = new Unit(100, UnitType.Percentage);
        container.Controls.Add(table);
        TableRow row = new TableRow();
        row.HorizontalAlign = HorizontalAlign.Right;
        row.Width = new Unit(100, UnitType.Percentage);
        table.Rows.Add(row);

        TableCell cell = new TableCell();
        cell.Width = new Unit(100, UnitType.Percentage);
        row.Cells.Add(cell);
        ASPxButton btn = new ASPxButton();
        btn.AutoPostBack = false;
        btn.Text = "Close";
        btn.ClientSideEvents.Click = string.Format("function(s, e){{ {0}.HideDropDown(); grid.AutoFilterByColumn('{0}', {0}.GetText()); }}", colName);
        cell.Controls.Add(btn);
    }
}

File: VB/WebSite/App_Code/DateTemplate.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Data
Imports System.Configuration
Imports System.Web
Imports System.Web.Security
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Web.UI.WebControls.WebParts
Imports System.Web.UI.HtmlControls
Imports DevExpress.Web.ASPxGridView
Imports DevExpress.Web.ASPxEditors


Public Class DataTemplate
	Implements ITemplate
	Private accessDS As AccessDataSource

	Public Sub New(ByVal accessDS As AccessDataSource)
		Me.accessDS = accessDS
	End Sub

	Public Sub InstantiateIn(ByVal container As Control) Implements ITemplate.InstantiateIn
		Dim dde As ASPxDropDownEdit = TryCast(container.NamingContainer.NamingContainer, ASPxDropDownEdit)
		Dim colName As String = dde.ClientInstanceName

		Dim listBox As New ASPxListBox()
		container.Controls.Add(listBox)

		listBox.ID = "listBox"
		listBox.Width = New Unit(100, UnitType.Percentage)
		listBox.ClientInstanceName = "cls" & colName
		listBox.SelectionMode = ListEditSelectionMode.CheckColumn
		listBox.ClientSideEvents.SelectedIndexChanged = String.Format("function(s,e) {{ OnListBoxSelectionChanged(s,e,{0}); }}", colName)

		Dim command As String = String.Format("SELECT DISTINCT [{0}] FROM [Categories]", colName)
		Dim ds As New AccessDataSource(accessDS.DataFile, command)

		Dim dv As DataView = TryCast(ds.Select(DataSourceSelectArguments.Empty), DataView)
		listBox.Items.Add("(Select all)")
		For i As Integer = 0 To dv.Count - 1
			listBox.Items.Add(dv(i)(0).ToString(),i)
		Next i

		Dim table As New Table()
		table.Width = New Unit(100, UnitType.Percentage)
		container.Controls.Add(table)
		Dim row As New TableRow()
		row.HorizontalAlign = HorizontalAlign.Right
		row.Width = New Unit(100, UnitType.Percentage)
		table.Rows.Add(row)

		Dim cell As New TableCell()
		cell.Width = New Unit(100, UnitType.Percentage)
		row.Cells.Add(cell)
		Dim btn As New ASPxButton()
		btn.AutoPostBack = False
		btn.Text = "Close"
		btn.ClientSideEvents.Click = String.Format("function(s, e){{ {0}.HideDropDown(); grid.AutoFilterByColumn('{0}', {0}.GetText()); }}", colName)
		cell.Controls.Add(btn)
	End Sub
End Class

